File: benchmarks/server/src/sirannon_bench/core/loadgen.py
"""Open-loop load generation with coordinated-omission correction.

A closed-loop generator sends the next request only after the previous one returns, so when the
server stalls the generator stops sending and the slow requests never get measured. That
undercount can hide the tail by orders of magnitude, which is the single failure a database
reviewer attacks first. This generator instead fires requests at a fixed target rate regardless
of whether earlier requests have returned, and it records each request's latency from the time
it was *meant* to be sent, not the time a backlog let it start. A request delayed by a stall
therefore carries the full delay, which is the wrk2 correction for coordinated omission.

Concurrency is bounded by a semaphore so a stalled server cannot spawn unbounded work. When the
semaphore blocks, the intended-time measurement still charges the wait to the delayed requests,
so the correction holds while memory stays bounded. The result records whether the target rate
was sustained and, when it was not, whether the bound was the server (the in-flight cap was hit)
or the driver itself (the cap was never reached), so a reader can tell a real result from a
generator that ran out of headroom.
"""

from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable
from dataclasses import dataclass

from .stats import mean, percentile


@dataclass(frozen=True)
class LoadResult:
    target_rate: float
    achieved_rate: float
    completed: int
    errors: int
    error_rate: float
    p50_ms: float
    p95_ms: float
    p99_ms: float
    p999_ms: float
    max_ms: float
    mean_ms: float
    max_in_flight_observed: int
    sustained: bool
    server_saturated: bool
    client_bound: bool


_SUSTAINED_FRACTION = 0.95


async def run_open_loop(
    run_op: Callable[[], Awaitable[bool]],
    target_rate: float,
    warmup_seconds: float,
    measure_seconds: float,
    max_in_flight: int,
) -> LoadResult:
    """Drive ``run_op`` at ``target_rate`` requests per second.

    ``run_op`` performs one operation and returns ``True`` on success or ``False`` on an
    application error; it must not raise, so a single failed request cannot tear down the run.
    Latencies from the warmup window are discarded; latencies from the measurement window are
    corrected for coordinated omission and returned as percentiles.
    """
    loop = asyncio.get_running_loop()
    semaphore = asyncio.Semaphore(max_in_flight)
    interval = 1.0 / target_rate

    start = loop.time()
    warmup_end = start + warmup_seconds
    measure_end = warmup_end + measure_seconds

    latencies_ms: list[float] = []
    served = 0
    errors = 0
    in_flight = 0
    max_in_flight_observed = 0
    tasks: set[asyncio.Task] = set()

    def in_measure_window(instant: float) -> bool:
        return warmup_end <= instant <= measure_end

    async def fire(intended_time: float, intended_in_window: bool) -> None:
        # Two accounting rules run here on purpose. Throughput counts responses that arrive within
        # the measurement window, so a saturated server whose backlog drains after the window is
        # not credited for work it did late. Latency is charged from the intended send time for
        # every request meant to be sent in the window, so a stalled request carries its full
        # delay. That split is the coordinated-omission correction.
        nonlocal served, errors, in_flight, max_in_flight_observed
        in_flight += 1
        max_in_flight_observed = max(max_in_flight_observed, in_flight)
        try:
            ok = await run_op()
        except asyncio.CancelledError:
            raise
        except Exception:
            ok = False
        finally:
            in_flight -= 1
            semaphore.release()
        arrival = loop.time()
        if intended_in_window and ok:
            latencies_ms.append((arrival - intended_time) * 1000.0)
        if in_measure_window(arrival):
            if ok:
                served += 1
            else:
                errors += 1

    index = 0
    while True:
        intended = start + index * interval
        if intended > measure_end:
            break
        delay = intended - loop.time()
        if delay > 0:
            await asyncio.sleep(delay)
        await semaphore.acquire()
        task = loop.create_task(fire(intended, intended >= warmup_end))
        tasks.add(task)
        task.add_done_callback(tasks.discard)
        index += 1

    # Wait for the backlog to drain, but not forever: at a saturating rate the tail would otherwise
    # take far longer than the run itself. Requests still outstanding after a grace window are
    # cancelled, which the saturation flag already reflects.
    if tasks:
        _, pending = await asyncio.wait(tasks, timeout=measure_seconds)
        for task in pending:
            task.cancel()
        await asyncio.gather(*tasks, return_exceptions=True)

    achieved_rate = served / measure_seconds if measure_seconds > 0 else 0.0
    total = served + errors
    error_rate = errors / total if total > 0 else 0.0
    sustained = achieved_rate >= target_rate * _SUSTAINED_FRACTION
    reached_cap = max_in_flight_observed >= max_in_flight
    return LoadResult(
        target_rate=target_rate,
        achieved_rate=achieved_rate,
        completed=served,
        errors=errors,
        error_rate=error_rate,
        p50_ms=percentile(latencies_ms, 0.50),
        p95_ms=percentile(latencies_ms, 0.95),
        p99_ms=percentile(latencies_ms, 0.99),
        p999_ms=percentile(latencies_ms, 0.999),
        max_ms=max(latencies_ms) if latencies_ms else 0.0,
        mean_ms=mean(latencies_ms),
        max_in_flight_observed=max_in_flight_observed,
        sustained=sustained,
        server_saturated=(not sustained) and reached_cap,
        client_bound=(not sustained) and not reached_cap,
    )

File: benchmarks/server/driver/src/loadgen.ts
import { type FailureCategory, FailureTally, OPERATION_REJECTED } from './failures.ts'
import { maxOf, mean, percentile } from './stats.ts'

export type RunOp = () => Promise<FailureCategory | null>

export interface LoadBucket {
  startSeconds: number
  served: number
  errors: number
  achievedRate: number
  p50Ms: number
  p99Ms: number
  maxMs: number
}

export interface LoadResult {
  targetRate: number
  achievedRate: number
  completed: number
  errors: number
  errorRate: number
  p50Ms: number
  p95Ms: number
  p99Ms: number
  p999Ms: number
  maxMs: number
  meanMs: number
  maxInFlight: number
  maxInFlightObserved: number
  capOccupancy: number
  occupancySamples: number
  offeredRate: number
  offeredFraction: number
  dispatchedInWindow: number
  expectedInWindow: number
  sustained: boolean
  reachedCap: boolean
  failures: FailureTally
  buckets: LoadBucket[] | null
}

const SUSTAINED_FRACTION = 0.95
const OCCUPANCY_SAMPLE_INTERVAL_MS = 10

class Semaphore {
  private permits: number
  private readonly waiters: Array<() => void> = []

  constructor(permits: number) {
    this.permits = permits
  }

  acquire(): Promise<void> {
    if (this.permits > 0) {
      this.permits -= 1
      return Promise.resolve()
    }
    return new Promise<void>(resolve => {
      this.waiters.push(resolve)
    })
  }

  release(): void {
    const waiter = this.waiters.shift()
    if (waiter) {
      waiter()
    } else {
      this.permits += 1
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise<void>(resolve => setTimeout(resolve, ms))
}

export async function runOpenLoop(
  runOp: RunOp,
  targetRate: number,
  warmupSeconds: number,
  measureSeconds: number,
  maxInFlight: number,
  bucketSeconds = 0,
): Promise<LoadResult> {
  const semaphore = new Semaphore(maxInFlight)
  const intervalMs = 1000 / targetRate

  const start = performance.now()
  const warmupEnd = start + warmupSeconds * 1000
  const measureEnd = warmupEnd + measureSeconds * 1000

  const bucketCount = bucketSeconds > 0 ? Math.ceil(measureSeconds / bucketSeconds) : 0
  const bucketServed = new Array<number>(bucketCount).fill(0)
  const bucketErrors = new Array<number>(bucketCount).fill(0)
  const latencyBucketIndex: number[] = []

  const latenciesMs: number[] = []
  const failures = new FailureTally()
  let served = 0
  let errors = 0
  let inFlight = 0
  let maxInFlightObserved = 0
  let dispatchedInWindow = 0
  const tasks = new Set<Promise<void>>()

  const inMeasureWindow = (instant: number): boolean => instant >= warmupEnd && instant <= measureEnd

  const fire = async (intendedTime: number, intendedInWindow: boolean): Promise<void> => {
    inFlight += 1
    if (inFlight > maxInFlightObserved) {
      maxInFlightObserved = inFlight
    }
    let failure: FailureCategory | null
    try {
      failure = await runOp()
    } catch {
      failure = OPERATION_REJECTED
    } finally {
      inFlight -= 1
      semaphore.release()
    }
    const arrival = performance.now()
    const bucket =
      bucketCount > 0 && intendedInWindow
        ? Math.min(bucketCount - 1, Math.floor((intendedTime - warmupEnd) / (bucketSeconds * 1000)))
        : -1
    if (intendedInWindow && failure === null) {
      latenciesMs.push(arrival - intendedTime)
      if (bucket >= 0) {
        latencyBucketIndex.push(bucket)
      }
    }
    if (bucket >= 0) {
      if (failure === null) {
        bucketServed[bucket] = (bucketServed[bucket] ?? 0) + 1
      } else {
        bucketErrors[bucket] = (bucketErrors[bucket] ?? 0) + 1
      }
    }
    if (inMeasureWindow(arrival)) {
      if (failure === null) {
        served += 1
      } else {
        errors += 1
        failures.record(failure)
      }
    }
  }

  let occupancySamples = 0
  let capSamples = 0
  const sampler = setInterval(() => {
    const instant = performance.now()
    if (!inMeasureWindow(instant)) {
      return
    }
    occupancySamples += 1
    if (inFlight >= maxInFlight) {
      capSamples += 1
    }
  }, OCCUPANCY_SAMPLE_INTERVAL_MS)
  sampler.unref()

  let dispatchEnd = warmupEnd
  try {
    let index = 0
    for (;;) {
      const intended = start + index * intervalMs
      if (intended > measureEnd) {
        break
      }
      const delay = intended - performance.now()
      if (delay > 0) {
        await sleep(delay)
      }
      await semaphore.acquire()
      const intendedInWindow = intended >= warmupEnd
      if (intendedInWindow) {
        dispatchedInWindow += 1
      }
      const task = fire(intended, intendedInWindow)
      tasks.add(task)
      void task.then(() => {
        tasks.delete(task)
      })
      index += 1
    }
    dispatchEnd = performance.now()

    await Promise.allSettled([...tasks])
  } finally {
    clearInterval(sampler)
  }

  const achievedRate = measureSeconds > 0 ? served / measureSeconds : 0.0
  const total = served + errors
  const errorRate = total > 0 ? errors / total : 0.0
  const sustained = achievedRate >= targetRate * SUSTAINED_FRACTION
  const reachedCap = maxInFlightObserved >= maxInFlight
  const expectedInWindow = Math.round(targetRate * measureSeconds)
  const offeredSpanSeconds = (dispatchEnd - warmupEnd) / 1000
  const offeredRate = offeredSpanSeconds > 0 ? dispatchedInWindow / offeredSpanSeconds : 0.0

  let buckets: LoadBucket[] | null = null
  if (bucketCount > 0) {
    const perBucket: number[][] = Array.from({ length: bucketCount }, () => [])
    for (const [position, bucketIdx] of latencyBucketIndex.entries()) {
      perBucket[bucketIdx]?.push(latenciesMs[position] ?? 0)
    }
    buckets = perBucket.map((samples, index) => ({
      startSeconds: index * bucketSeconds,
      served: bucketServed[index] ?? 0,
      errors: bucketErrors[index] ?? 0,
      achievedRate: (bucketServed[index] ?? 0) / Math.min(bucketSeconds, measureSeconds - index * bucketSeconds),
      p50Ms: percentile(samples, 0.5),
      p99Ms: percentile(samples, 0.99),
      maxMs: maxOf(samples),
    }))
  }

  return {
    targetRate,
    achievedRate,
    completed: served,
    errors,
    errorRate,
    p50Ms: percentile(latenciesMs, 0.5),
    p95Ms: percentile(latenciesMs, 0.95),
    p99Ms: percentile(latenciesMs, 0.99),
    p999Ms: percentile(latenciesMs, 0.999),
    maxMs: maxOf(latenciesMs),
    meanMs: mean(latenciesMs),
    maxInFlight,
    maxInFlightObserved,
    capOccupancy: occupancySamples > 0 ? capSamples / occupancySamples : 0.0,
    occupancySamples,
    offeredRate,
    offeredFraction: targetRate > 0 ? offeredRate / targetRate : 0.0,
    dispatchedInWindow,
    expectedInWindow,
    sustained,
    reachedCap,
    failures,
    buckets,
  }
}

export interface ClientCeiling {
  ceilingOps: number
  errors: number
  concurrency: number
  seconds: number
  statement: string
}

export async function measureClientCeiling(
  runOp: RunOp,
  seconds: number,
  concurrency: number,
  statement: string,
): Promise<ClientCeiling> {
  const deadline = performance.now() + seconds * 1000
  let ops = 0
  let errors = 0

  const worker = async (): Promise<void> => {
    while (performance.now() < deadline) {
      let failure: FailureCategory | null
      try {
        failure = await runOp()
      } catch {
        failure = OPERATION_REJECTED
      }
      if (failure === null) {
        ops += 1
      } else {
        errors += 1
      }
    }
  }

  const workers: Array<Promise<void>> = []
  for (let i = 0; i < concurrency; i++) {
    workers.push(worker())
  }
  await Promise.all(workers)

  return { ceilingOps: ops / seconds, errors, concurrency, seconds, statement }
}
